---
title: "Population stratification report"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
library(tidyverse)
knitr::opts_chunk$set(echo = TRUE)
```


```{r input_params, include=FALSE}
# Access params in RMarkdown
params <- list(
    db = params$db,
    fam = params$fam,
    cores = params$cores,
    TOPMed_imputation = params$TOPMed_imputation,
    pop_strat_filter = params$pop_strat_filter,
    vcf = params$vcf,
    bed = params$bed,
    bim = params$bim,
    results = params$results,
    convert_build = params$convert_build,
    qc = params$qc,
    pop_strat = params$pop_strat,
    impute = params$impute,
    pre_impute = params$pre_impute,
    output_chr = params$output_chr,
    post_impute = params$post_impute,
    gwas = params$gwas,
    gwas_lmm = params$gwas_lmm,
    list_final_patients = params$list_final_patients,
    download_db = params$download_db,
    help = params$help,
    input_build = params$input_build,
    output_build = params$output_build,
    mem = params$mem,
    sexcheck = params$sexcheck,
    F_threshold_male = params$F_threshold_male,
    F_threshold_female = params$F_threshold_female,
    keep_sex_chroms = params$keep_sex_chroms,
    mind = params$mind,
    indep_pairwise = params$indep_pairwise,
    bad_ld = params$bad_ld,
    variant_geno = params$variant_geno,
    king_cutoff = params$king_cutoff,
    hwe = params$hwe,
    maf = params$maf,
    missingness = params$missingness,
    popfile = params$popfile,
    popcode = params$popcode,
    parfile = params$parfile,
    pca_covars = params$pca_covars,
    rm_missing_pheno = params$rm_missing_pheno,
    heterozygosity = params$heterozygosity,
    covar_file = params$covar_file,
    linear = params$linear,
    topmed_password = params$topmed_password,
    impute_chroms = params$impute_chroms,
    info = params$info,
    r2 = params$r2,
    impute_maf = params$impute_maf,
    qc_fam = params$qc_fam,
    impute5_version = params$impute5_version
)
```


The aim of the population stratification is to identify individuals of an ancestry other than the reported one, based on population structure inferred from 1,000 Genomes Project. This step minimises the possibility that the difference in the alleles frequencies is caused by different ancestry of the samples. 

## PCA 

Principal Components Analysis (PCA) is performed using those SNPs shared between our sample (marked as 'OWN') and the individuals in 1,000 Genomes Project. Subjects of the same ancestry are expected to cluster together. It is recommended to exclude those subjects whose reported and inferred ancestry are not the same. However, if you wish to impute SNPs for all subjects, you can exclude such patients later. 

In this analysis, the parameter `pop_strat_filter` is set to `r params$pop_strat_filter`.

### PC1 vs PC2

#### Before outlier removal

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("PC1vsPC2_before.png")
```

#### After outlier removal

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("PC1vsPC2_after.png")
```

### PC1 vs PC3

#### Before outlier removal

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("PC1vsPC3_before.png")
```

#### After outlier removal

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("PC1vsPC3_after.png")
```

### PC2 vs PC3

#### Before outlier removal

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("PC2vsPC3_before.png")
```

#### After outlier removal

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("PC2vsPC3_after.png")
```

### 3D

#### Before outlier removal

```{r, echo=FALSE, out.width="80%"}
pca_before <- readRDS("plink_3D_pcabefore.rds")
pca_before
```

#### After outlier removal

```{r, echo=FALSE, out.width="80%"}
pca_after <- readRDS("plink_3D_pcaafter.rds")
pca_after
```

## Log

```{r, echo=FALSE, out.width="80%"}
knitr::include_graphics("pop_strat_log_samples.png")
knitr::include_graphics("pop_strat_log_variants.png")
```
